src/lib/rnp2.cpp

```diff
@@ -3226,180 +3294,775 @@ rnp_generate_key_json(rnp_ffi_t ffi, const char *json, char **results)
+        op->crypto.elgamal.key_bitlen = bits;
+        break;
+    case PGP_PKA_DSA:
+        op->crypto.dsa.p_bitlen = bits;
+        break;
+    default:
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+
+    return RNP_SUCCESS;
 }
 
 rnp_result_t
-rnp_key_get_uid_count(rnp_key_handle_t handle, size_t *count)
+rnp_op_generate_set_hash(rnp_op_generate_t op, const char *hash)
 {
-    if (handle == NULL || count == NULL) {
+    if (!op || !hash) {
         return RNP_ERROR_NULL_POINTER;
     }
-
-    pgp_key_t *key = get_key_prefer_public(handle);
-    *count = pgp_key_get_userid_count(key);
+    if (!str_to_hash_alg(hash, &op->crypto.hash_alg)) {
+        FFI_LOG(op->ffi, "Invalid hash: %s", hash);
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_generate_set_dsa_qbits(rnp_op_generate_t op, uint32_t qbits)
+{
+    if (!op) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+    if (op->crypto.key_alg != PGP_PKA_DSA) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    op->crypto.dsa.q_bitlen = qbits;
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_generate_set_curve(rnp_op_generate_t op, const char *curve)
+{
+    if (!op || !curve) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+    if (!pk_alg_allows_custom_curve(op->crypto.key_alg)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    if (!curve_str_to_type(curve, &op->crypto.ecc.curve)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_generate_set_protection_cipher(rnp_op_generate_t op, const char *cipher)
+{
+    if (!op || !cipher) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+    if (!str_to_cipher(cipher, &op->protection.symm_alg)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_generate_set_protection_hash(rnp_op_generate_t op, const char *hash)
+{
+    if (!op || !hash) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+    if (!str_to_hash_alg(hash, &op->protection.hash_alg)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_generate_set_protection_mode(rnp_op_generate_t op, const char *mode)
+{
+    if (!op || !mode) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+    if (!str_to_cipher_mode(mode, &op->protection.cipher_mode)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_generate_set_protection_iterations(rnp_op_generate_t op, uint32_t iterations)
+{
+    if (!op) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+    op->protection.iterations = iterations;
+    return RNP_SUCCESS;
+}
+
+rnp_result_t
+rnp_op_generate_add_usage(rnp_op_generate_t op, const char *usage)
+{
+    if (!op || !usage) {
+        return RNP_ERROR_NULL_POINTER;
+    }
+    uint8_t flag = 0;
+    if (!str_to_key_flag(usage, &flag)) {
+        return RNP_ERROR_BAD_PARAMETERS;
+    }
+    if (op->primary) {
```

Comment:
I'm wondering if we ought to do a check with `pgp_pk_alg_capabilities` here. What do you think?

Reply:
@dewyatt Yeah, correct. Just pushed a commit with some tests.